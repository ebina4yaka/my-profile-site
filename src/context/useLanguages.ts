import { createContext, useCallback, useState } from 'react'
import { Language } from '../lib/models'

export type LanguagesContext = {
  languages: Language[]
  setLanguages: (languages: Language[]) => void
}

const defaultContext: LanguagesContext = {
  languages: [],
  setLanguages: () => {},
}

export const languagesContext = createContext<LanguagesContext>(defaultContext)

export default function useLanguages(): LanguagesContext {
  const [languages, setLanguagesState] = useState<Language[]>([])
  const setLanguages = useCallback(
    (current: Language[]): void => setLanguagesState(current),
    []
  )

  return { languages, setLanguages }
}
