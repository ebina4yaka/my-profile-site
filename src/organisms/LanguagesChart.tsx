import { ReactElement, useContext } from 'react'
import { Cell, Pie, PieChart } from 'recharts'
import { languagesContext } from '../context/useLanguages'

export default function LanguagesChart(): ReactElement {
  const { languages } = useContext(languagesContext)
  return (
    <div className="container">
      <div className="sm:box-content py-12">
        <div className="font-sans font-bold text-3xl text-center pb-2">
          Languages Chart
        </div>
        <div className="flex flex-row flex-wrap justify-center items-center content-center">
          <PieChart id="userLanguages" width={180} height={180}>
            <Pie
              data={languages}
              dataKey="percentage"
              nameKey="name"
              cx="50%"
              cy="50%"
            >
              {languages.map((language) => (
                <Cell
                  key={`cells-${language.name}`}
                  fill={language.color}
                  strokeWidth={0}
                />
              ))}
            </Pie>
          </PieChart>
          <div className="flex flex-col flex-wrap justify-center content-center">
            {languages.map((language) => {
              return (
                <div
                  key={language.name}
                  className="flex flex-row justify-start items-center"
                >
                  <span
                    className="h-3 w-3 rounded-full inline-block mr-1 mt-0.5"
                    style={{ backgroundColor: language.color }}
                  />
                  <span className="text-base font-mono">
                    {`${language.name} ${language.percentage}%`}
                  </span>
                </div>
              )
            })}
          </div>
        </div>
      </div>
    </div>
  )
}
