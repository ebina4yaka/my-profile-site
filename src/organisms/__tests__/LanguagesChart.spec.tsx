import renderer from 'react-test-renderer'
import { act, renderHook } from '@testing-library/react-hooks'
import LanguagesChart from '../LanguagesChart'
import useLanguages, { languagesContext } from '../../context/useLanguages'

it('Snapshot test for LanguagesChart', () => {
  const { result } = renderHook(() => useLanguages())
  const { setLanguages } = result.current
  act(() => {
    setLanguages([
      {
        name: 'Rust',
        color: '#dea584',
        percentage: 25,
      },
      {
        name: 'Go',
        color: '#00ADD8',
        percentage: 25,
      },
      {
        name: 'Dart',
        color: '#00B4AB',
        percentage: 25,
      },
      {
        name: 'Python',
        color: '#3572A5',
        percentage: 25,
      },
    ])
  })

  const tree = renderer
    .create(
      <languagesContext.Provider value={result.current}>
        <LanguagesChart />
      </languagesContext.Provider>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
