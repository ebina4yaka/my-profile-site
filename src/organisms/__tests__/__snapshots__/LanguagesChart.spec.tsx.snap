// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test for LanguagesChart 1`] = `
<div
  className="container"
>
  <div
    className="sm:box-content py-12"
  >
    <div
      className="font-sans font-bold text-3xl text-center pb-2"
    >
      Languages Chart
    </div>
    <div
      className="flex flex-row flex-wrap justify-center items-center content-center"
    >
      <div
        className="recharts-wrapper"
        style={
          Object {
            "cursor": "default",
            "height": 180,
            "position": "relative",
            "width": 180,
          }
        }
      >
        <svg
          className="recharts-surface"
          cx="50%"
          cy="50%"
          height={180}
          id="userLanguages"
          version="1.1"
          viewBox="0 0 180 180"
          width={180}
        >
          <defs>
            <clipPath
              id="userLanguages-clip"
            >
              <rect
                height={170}
                width={170}
                x={5}
                y={5}
              />
            </clipPath>
          </defs>
          <g
            className="recharts-layer recharts-pie"
          >
            <g
              className="recharts-layer"
            >
              <g
                className="recharts-layer recharts-pie-sector"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              />
              <g
                className="recharts-layer recharts-pie-sector"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              />
              <g
                className="recharts-layer recharts-pie-sector"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              />
              <g
                className="recharts-layer recharts-pie-sector"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              />
            </g>
          </g>
        </svg>
      </div>
      <div
        className="flex flex-col flex-wrap justify-center content-center"
      >
        <div
          className="flex flex-row justify-start items-center"
        >
          <span
            className="h-3 w-3 rounded-full inline-block mr-1 mt-0.5"
            style={
              Object {
                "backgroundColor": "#dea584",
              }
            }
          />
          <span
            className="text-base font-mono"
          >
            Rust 25%
          </span>
        </div>
        <div
          className="flex flex-row justify-start items-center"
        >
          <span
            className="h-3 w-3 rounded-full inline-block mr-1 mt-0.5"
            style={
              Object {
                "backgroundColor": "#00ADD8",
              }
            }
          />
          <span
            className="text-base font-mono"
          >
            Go 25%
          </span>
        </div>
        <div
          className="flex flex-row justify-start items-center"
        >
          <span
            className="h-3 w-3 rounded-full inline-block mr-1 mt-0.5"
            style={
              Object {
                "backgroundColor": "#00B4AB",
              }
            }
          />
          <span
            className="text-base font-mono"
          >
            Dart 25%
          </span>
        </div>
        <div
          className="flex flex-row justify-start items-center"
        >
          <span
            className="h-3 w-3 rounded-full inline-block mr-1 mt-0.5"
            style={
              Object {
                "backgroundColor": "#3572A5",
              }
            }
          />
          <span
            className="text-base font-mono"
          >
            Python 25%
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
